### ISSUE TITLE
cf purge-service-offering should purge multiple offerings registered within same broker

### ISSUE THREAD

**Issue by gberche-orange** [cloudfoundry/cli#1859]
**Describe the bug and the command you saw an issue with**

As a CF admin, in order to purge multiple service offerings with same name registred within a single broker, I need cf purge-service-offering to purge at least one offering 

**What happened**

```
cf purge-service-offering p-mysql-cmdb -b osb-cmdb-broker-0 
Service 'p-mysql-cmdb' is provided by multiple service brokers. Specify a broker by using the '-b' flag.
FAILED
```

**Expected behavior**

```cf purge-service-offering p-mysql-cmdb -b osb-cmdb-broker-0 
Warning service 'p-mysql-cmdb' is registered multiple times with broker osb-cmdb-broker-0, purging first instance with guid 2b991d81-730e-42f3-9f7e-37b375a1944e
...
OK
```

**To Reproduce**
Steps to reproduce the behavior; include the exact CLI commands and verbose output:

given a service offering registered multiple times with the same broker, then run the commands above

**Provide more context**
- platform and shell details: bash, Ubuntu 18.04
- version of the CLI you are running: cf version 6.47.2+d526c2cb3.2019-11-05
- version of the CC API Release you are on Cf deployment 12

/CC @addytripathi 

**Workaround**

* Identify the guid of the 1st service offering to delete using `CF_TRACE=true cf purge-service-offering p-mysql-cmdb -b osb-cmdb-broker-0`
* maually purge the 1st offering `cf curl -X DELETE '/v2/services/2b991d81-730e-42f3-9f7e-37b375a1944e?purge=true'`
* Retry the cf purge command `cf purge-service-offering p-mysql-cmdb -b osb-cmdb-broker-0 `

> WARNING: This operation assumes that the service broker responsible for this service offering is no longer available, and all service instances have been deleted, leaving orphan records in Cloud Foundry's database. All knowledge of the service will be removed from Cloud Foundry, including service instances and service bindings. No attempt will be made to contact the service broker; running this command without destroying the service broker will cause orphan service instances. After running this command you may want to run either delete-service-auth-token or delete-service-broker to complete the cleanup.
> 
> Really purge service offering p-mysql-cmdb from broker osb-cmdb-broker-0 from Cloud Foundry? [yN]: y
> Purging service p-mysql-cmdb...
> OK
> 

**debugging traces**

```
CF_TRACE=true cf purge-service-offering p-mysql-cmdb -b osb-cmdb-broker-0 | grep 'GET '
GET /v2/info HTTP/1.1
GET /login HTTP/1.1
GET /v2/service_brokers?q=name%3Aosb-cmdb-broker-0 HTTP/1.1
GET /v2/services?q=label%3Ap-mysql-cmdb&q=service_broker_guid%3Ae3ba8c1c-76ab-4da4-8973-3dd5b2126e68 HTTP/1.1
Service 'p-mysql-cmdb' is provided by multiple service brokers. Specify a broker by using the '-b' flag.
```

```
REQUEST: [2019-12-03T11:55:43+01:00]
GET /v2/service_brokers?q=name%3Aosb-cmdb-broker-0 HTTP/1.1
Host: redacted
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
User-Agent: cf/6.47.2+d526c2cb3.2019-11-05 (go1.12.12; amd64 linux)

RESPONSE: [2019-12-03T11:55:43+01:00]
HTTP/1.1 200 OK
Content-Length: 609
Content-Type: application/json;charset=utf-8
Date: Tue, 03 Dec 2019 10:55:42 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 8876033e-a494-49ef-7fbe-7497ea16f34b::8f81ac7d-2a95-4217-8468-02e5247f95dd
{
  "next_url": null,
  "prev_url": null,
  "resources": [
    {
      "entity": {
        "auth_username": "missing-broker-username-from-secrets-file",
        "broker_url": "https://osb-cmdb-broker-0.redacted.org",
        "name": "osb-cmdb-broker-0",
        "space_guid": null
      },
      "metadata": {
        "created_at": "2019-11-18T21:45:31Z",
        "guid": "e3ba8c1c-76ab-4da4-8973-3dd5b2126e68",
        "updated_at": "2019-12-03T10:21:24Z",
        "url": "/v2/service_brokers/e3ba8c1c-76ab-4da4-8973-3dd5b2126e68"
      }
    }
  ],
  "total_pages": 1,
  "total_results": 1
}


REQUEST: [2019-12-03T11:55:43+01:00]
GET /v2/services?q=label%3Ap-mysql-cmdb&q=service_broker_guid%3Ae3ba8c1c-76ab-4da4-8973-3dd5b2126e68 HTTP/1.1
Host: redacted
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
User-Agent: cf/6.47.2+d526c2cb3.2019-11-05 (go1.12.12; amd64 linux)

RESPONSE: [2019-12-03T11:55:43+01:00]
HTTP/1.1 200 OK
Content-Length: 6894
Content-Type: application/json;charset=utf-8
Date: Tue, 03 Dec 2019 10:55:42 GMT
Server: nginx
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 3633c4d2-deec-418d-42b0-9791c9d2d5be::02e3eac7-f273-4ad5-8216-c52d976f1165
{
  "next_url": null,
  "prev_url": null,
  "resources": [
    {
      "entity": {
        "active": false,
        "allow_context_updates": false,
        "bindable": true,
        "bindings_retrievable": false,
        "description": "A useful service",
        "documentation_url": null,
        "extra": "{\n\n}",
        "info_url": null,
        "instances_retrievable": false,
        "label": "p-mysql-cmdb",
        "long_description": null,
        "plan_updateable": true,
        "provider": null,
        "requires": null,
        "service_broker_guid": "e3ba8c1c-76ab-4da4-8973-3dd5b2126e68",
        "service_broker_name": "osb-cmdb-broker-0",
        "service_plans_url": "/v2/services/2b991d81-730e-42f3-9f7e-37b375a1944e/service_plans",
        "tags": [
          "example"
        ],
        "unique_id": "ebca66fd-461d-415b-bba3-5e379d671c88",
        "url": null,
        "version": null
      },
      "metadata": {
        "created_at": "2019-11-18T21:45:31Z",
        "guid": "2b991d81-730e-42f3-9f7e-37b375a1944e",
        "updated_at": "2019-12-03T10:21:24Z",
        "url": "/v2/services/2b991d81-730e-42f3-9f7e-37b375a1944e"
      }
    },
    {
      "entity": {
        "active": true,
        "allow_context_updates": false,
        "bindable": true,
        "bindings_retrievable": false,
        "description": "MariaDB databases 10.1.38 on demand on shared cluster",
        "documentation_url": null,
        "extra": "{\n  \"longDescription\": ... /\"\n}",
        "info_url": null,
        "instances_retrievable": false,
        "label": "p-mysql-cmdb",
        "long_description": null,
        "plan_updateable": true,
        "provider": null,
        "requires": null,
        "service_broker_guid": "e3ba8c1c-76ab-4da4-8973-3dd5b2126e68",
        "service_broker_name": "osb-cmdb-broker-0",
        "service_plans_url": "/v2/services/8e4ad2ce-e589-41ab-b5ce-361acd6c8321/service_plans",
        "tags": [
          "mysql"
        ],
        "unique_id": "22450290-aa7f-4349-9100-92803ce9b3b3",
        "url": null,
        "version": null
      },
      "metadata": {
        "created_at": "2019-12-03T10:21:24Z",
        "guid": "8e4ad2ce-e589-41ab-b5ce-361acd6c8321",
        "updated_at": "2019-12-03T10:21:24Z",
        "url": "/v2/services/8e4ad2ce-e589-41ab-b5ce-361acd6c8321"
      }
    }
  ],
  "total_pages": 1,
  "total_results": 2
}
```

**Comment by FelisiaM** [comment]
Hi @gberche-orange 

Thanks for opening this and apologies for the very late response.

I attempted the reproduce your issue but I can no longer recreate it using CLI v7 backed by v3 CAPI.
I tried registering a broker with a catalog that has the same service offering name twice. This caused create/update operations to fail when checking the catalog with error:
```
Service broker catalog is invalid:
Service names must be unique within a broker
```

This is what I expected to happen. 

[Since 2018](https://github.com/cloudfoundry/cloud_controller_ng/commit/9def95f3526483de6e006f81e19a9e32804964c9#diff-6a0d6573c6383e5a99c0491d53ccc500dad6f0cfab3aebceee1994d8a66aadc9) you could not register a broker containing the same offering name twice. In the latest [v3 API](https://github.com/cloudfoundry/cloud_controller_ng/commit/f7134bd67bb15fd94dbdb74ad6426789b3596f4d#diff-6a0d6573c6383e5a99c0491d53ccc500dad6f0cfab3aebceee1994d8a66aadc9) we have also made sure that even when you are updating a broker the operation does not result in registering the same offering name twice.

Given that this cannot happen anymore and there is a workaround for any older brokers I will close the issue.

Thanks,
Felisia

**Comment by gberche-orange** [comment]
thanks @FelisiaM for the time to looking into this issue and try to reproduce, and sorry my report was incorrect/incomplete in particular the `how to reproduce`.

Double checking [cf-deployment-12](https://github.com/cloudfoundry/cf-deployment/blob/50cfcf63e94b678a932516c277ece755cf1a140d/cf-deployment.yml#L2370-L2372) with [capi 1.92.0](https://github.com/cloudfoundry/capi-release/tree/1.92.0/src
https://github.com/cloudfoundry/cloud_controller_ng/tree/12a35f78822ce7600b381b4c796eb429d1c6e7b8) indeed the catalog [is being validated](
https://github.com/cloudfoundry/cloud_controller_ng/blob/12a35f78822ce7600b381b4c796eb429d1c6e7b8/lib/services/service_brokers/v2/catalog.rb) before loading it.

I'll try to reproduce the issue with the following scenario in v3:
1- create a new service broker when its v2/catalog returns a single service name="name" id=guid1, with plan id=guid2
2- update service broker when its v2/catalog returns a different single service name="name" id=guid3, with plan id=guid4
3- try to delete services with name "name": cf CLI fails as reported

I suspect step 2) was accepted with v2 because 
* the candidate catalog is valid at that time (i.e. `validate_catalog!`) because no duplicate in names or ids
* the merged existing catalog and new catalog (which has duplicate services with same name and distinct ids) isn't rejected by `synchronize_services_and_plans!` method (see code extracts below)

In between, if ever your knowledge of v3 implementation confirms this scenario would still be possible with v3, that'd be quite useful as well.

https://github.com/cloudfoundry/cloud_controller_ng/blob/12a35f78822ce7600b381b4c796eb429d1c6e7b8/lib/services/service_brokers/service_broker_registration.rb#L43-L50

```ruby
      validate_catalog!
      route_service_warning unless @route_services_enabled
      volume_mount_service_warning unless @volume_services_enabled
      synchronize_dashboard_clients!

      broker.db.transaction do
        broker.save
        synchronize_services_and_plans!
```

```ruby
    def synchronize_services_and_plans!
      @service_manager.sync_services_and_plans(catalog)

      if @service_manager.has_warnings?
        @service_manager.warnings.each { |warning| warnings << warning }
      end
    end
```

https://github.com/cloudfoundry/cloud_controller_ng/blob/12a35f78822ce7600b381b4c796eb429d1c6e7b8/lib/services/service_brokers/service_manager.rb#L11-L18

```ruby
    def sync_services_and_plans(catalog)
      update_or_create_services(catalog)
      deactivate_services(catalog)
      update_or_create_plans(catalog)
      deactivate_plans(catalog)
      delete_plans(catalog)
      delete_services(catalog)
    end
```

https://github.com/cloudfoundry/cloud_controller_ng/blob/12a35f78822ce7600b381b4c796eb429d1c6e7b8/lib/services/service_brokers/service_manager.rb#L26-L50

```ruby
    def update_or_create_services(catalog)
      existing_services, new_services = catalog.services.partition do |service|
        VCAP::CloudController::Service.where(
          service_broker: service.service_broker,
          unique_id: service.broker_provided_id
        ).present?
      end

      existing_services.each do |catalog_service|
        cond = {
          service_broker: catalog_service.service_broker,
          unique_id:      catalog_service.broker_provided_id,
        }
        service = VCAP::CloudController::Service.find(cond)
        update_service_from_catalog(service, catalog_service)
      end

      new_services.each do |catalog_service|
        service = VCAP::CloudController::Service.new(
          unique_id: catalog_service.broker_provided_id,
          service_broker: catalog_service.service_broker,
        )
        update_service_from_catalog(service, catalog_service)
      end
    end
```

**Comment by gberche-orange** [comment]
Edited scenario with step 1 `create service instance`

**Comment by FelisiaM** [comment]
Thanks for the update @gberche-orange 

I believe you are describing this issue https://www.pivotaltracker.com/n/projects/2105761/stories/172672538 and I can confirm that it has been resolved in v3 capi release [v1.95.0](https://github.com/cloudfoundry/capi-release/releases/tag/1.95.0) included in cf-deployment [v13.5.0](https://github.com/cloudfoundry/cf-deployment/releases/tag/v13.5.0). 

Let us know if you find any issues.
